clang/lib/CIR/Dialect/Transforms/LoweringPrepareAArch64CXXABI.cpp

```diff
@@ -0,0 +1,176 @@
+  // registers or not.
+  auto maybeRegCmpRes = builder.create<mlir::cir::CmpOp>(
+      loc, boolTy, CmpOpKind::le, newRegsOffs, zeroValue);
+  builder.create<mlir::cir::BrCondOp>(loc, maybeRegCmpRes, inRegBlock,
+                                      onStackBlock);
+
+  //=======================================
+  // Argument was on the stack
+  //=======================================
+  builder.setInsertionPointToEnd(onStackBlock);
+  auto stackP = builder.createGetMemberOp(loc, valist, "stack", 0);
+  auto stack = builder.create<mlir::cir::LoadOp>(loc, stackP);
+  auto ptrDiffTy =
+      mlir::cir::IntType::get(builder.getContext(), 64, /*signed=*/false);
+  auto eight = builder.create<mlir::cir::ConstantOp>(
+      loc, ptrDiffTy, mlir::cir::IntAttr::get(ptrDiffTy, 8));
+  auto i8Ty = IntegerType::get(builder.getContext(), 8);
+  auto i8PtrTy = PointerType::get(builder.getContext(), i8Ty);
+  auto castStack = builder.createBitcast(stack, i8PtrTy);
+  // Write the new value of __stack for the next call to va_arg
+  auto newStackAsi8Ptr = builder.create<mlir::cir::PtrStrideOp>(
+      loc, castStack.getType(), castStack, eight);
+  auto newStack = builder.createBitcast(newStackAsi8Ptr, stack.getType());
+  builder.createStore(loc, newStack, stackP);
+  builder.create<mlir::cir::BrOp>(loc, mlir::ValueRange{stack}, newEndBlock);
+
+  //=======================================
+  // Argument was in registers
+  //=======================================
+  // Now we emit the code for if the argument was originally passed in
+  // registers. First start the appropriate block:
+  builder.setInsertionPointToEnd(inRegBlock);
+  auto regTopP = builder.createGetMemberOp(loc, valist,
+                                           isFloatingType ? "vr_top" : "gr_top",
+                                           isFloatingType ? 2 : 1);
+  auto regTop = builder.create<mlir::cir::LoadOp>(loc, regTopP);
+  auto castRegTop = builder.createBitcast(regTop, i8PtrTy);
+  auto resAsInt8P = builder.create<mlir::cir::PtrStrideOp>(
+      loc, castRegTop.getType(), castRegTop, offs);
+  auto resAsVoidP = builder.createBitcast(resAsInt8P, regTop.getType());
+  builder.create<mlir::cir::BrOp>(loc, mlir::ValueRange{resAsVoidP},
+                                  newEndBlock);
+
+  // generate additional instructions for end block
+  builder.setInsertionPoint(op);
+  newEndBlock->addArgument(stack.getType(), loc);
+  auto resP = newEndBlock->getArgument(0);
+  assert(resP.getType().isa<mlir::cir::PointerType>());
+  auto opResPTy = PointerType::get(builder.getContext(), opResTy);
+  auto castResP = builder.createBitcast(resP, opResPTy);
+  auto res = builder.create<mlir::cir::LoadOp>(loc, castResP);
+  return res.getResult();
+}
+
+mlir::Value LoweringPrepareAArch64CXXABI::lowerVAArg(CIRBaseBuilderTy &builder,
+                                                     mlir::cir::VAArgOp op) {
+
+  if (Kind == clang::TargetCXXABI::GenericAArch64) {
```

Comment:
`AArch64ABIKind` should be used here.

Reply:
Unfortunately, TargetCXXABI has its own list of enum values defined in clang/include/clang/Basic/TargetCXXABI.def
and 
AArch64ABIKind is defined lib/CodeGen/TargetInfo.h, 

In clang/lib/Basic/Targets/AArch64.cpp
We link them together. other than AArch64ABIKind, 
only MinGWARM64TargetInfo is using TargetCXXABI::GenericAArch64.  Would that be a concern for us?

Reply:
> Unfortunately, TargetCXXABI has its own list of enum values defined in clang/include/clang/Basic/TargetCXXABI.def

Everything defined include/Basic should be already usable by CIRGen.

> and AArch64ABIKind is defined lib/CodeGen/TargetInfo.h,

CIRGen also has a incipient `clang/lib/CIR/CodeGen/TargetInfo.h`, please add `AArch64ABIKind` or any other machinery there.

> other than AArch64ABIKind, only MinGWARM64TargetInfo is using TargetCXXABI::GenericAArch64. Would that be a concern for us?

Note sure what you mean here, but we need to think a further than our specific/immediate needs and leave room for open source to easily collaborate on top of things we add - this is why we add asserts and try to mimic the OG skeleton.

Reply:
got is, kind should be TargetInfo not TargetCXXABI.